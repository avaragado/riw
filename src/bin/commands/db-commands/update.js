// @flow

import type yargs from 'yargs';

import outdent from 'outdent';
import pick from 'ramda/src/pick';

import type { RIW, DBUpdateSpec } from '../../..';
import { createHandlerWithRIW } from '../../utils';

export const command = 'update';
export const desc = 'Add or edit a translation in the riw database';

const here = `db ${command}`;

export const builder = (yyargs: yargs.Argv) => yyargs
    .usage(outdent`
        $0 ${here}
            --defaultMessage <sourceString> [--description <description>]
            --locale <localeID> --translation <localeString>

        Update the translation for <sourceString> (optionally disambiguated by <description>)
        in the default locale to <localeString> in the <localeID> locale.
    `)
    .example(
        outdent`
            $0 ${here}
                --defaultMessage "Hello {name}"
                --locale fr-FR
                --translation "Bonjour {name}"

        `,
        'Update the fr-FR translation for a welcome message.',
    )
    .example(
        outdent`
            $0 ${here}
                --defaultMessage "Export"
                --description "The button a user presses to start exporting data"
                --locale fr-FR
                --translation "Exporter"

        `,
        'Update the fr-FR translation for a button, with a description to disambiguate meaning.',
    )
    .example(
        outdent`
            $0 ${here}
                --defaultMessage "Export"
                --description "The section heading describing an export"
                --locale fr-FR
                --translation "Exportation"

        `,
        'Update the fr-FR translation for a label, with a description to disambiguate meaning.',
    )
    .example(
        outdent`
            $0 ${here}
                --defaultMessage "Hello"
                --description 'JSON:{"property": "value", "property2": 123}'
                --locale pt-BR
                --translation "OlÃ¡"
        `,
        'Update a pt-BR translation, using an object description expressed as a JSON string.',
    )
    .options({
        defaultMessage: {
            alias: 'm',
            string: true,
            demandOption: true,
            group: 'Command options',
            desc: outdent`
                The string that's being translated, in the default locale.
            `,
        },

        description: {
            alias: 'd',
            string: true,
            group: 'Command options',
            desc: outdent`
                The description for the string being translated.
                Required IF you use the 'description' field in react-intl message descriptors,
                AND if you use the same 'defaultMessage' in several places,
                AND you need different translations for these identical defaultMessages.
                If this has the prefix "JSON:", the remainder is parsed as JSON. Otherwise it's left as-is.
            `,
            coerce: (arg: string) => (arg.startsWith('JSON:') ? JSON.parse(arg.slice(5)) : arg),
        },

        locale: {
            alias: 'l',
            string: true,
            demandOption: true,
            group: 'Command options',
            desc: outdent`
                The locale ID of the translated string.
            `,
        },

        translation: {
            alias: 't',
            string: true,
            demandOption: true,
            group: 'Command options',
            desc: outdent`
                The translated string.
            `,
        },
    });

export const handler = createHandlerWithRIW((riw: RIW, argv: yargs.Argv) => {
    const opt: DBUpdateSpec = {
        translations: [
            pick(['defaultMessage', 'description', 'locale', 'translation'], argv),
        ],
    };

    try {
        riw.db.update(opt);

    } catch (err) {
        // ignore
    }
});
