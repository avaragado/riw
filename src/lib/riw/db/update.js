// @flow

import compose from 'ramda/src/compose';

import type {
    TranslatedMessageDescriptor,
} from '../../../types';
import type { Config } from '../../config';

import dbRead from './rw/dbRead';
import writeFromDB from './rw/writeFromDB';
import arquadFromDB from './transform/arquadFromDB';
import dbFromQuadAr from './transform/dbFromQuadAr';
import arquadUpdateFromQuadAr from './transform/arquadUpdateFromQuadAr';

export type DBUpdateSpec = {
    translations: TranslatedMessageDescriptor[],
};

export default (config: Config) => (opt?: DBUpdateSpec) => compose(
    writeFromDB(config),
    dbFromQuadAr,
    arquadUpdateFromQuadAr(opt),
    arquadFromDB,
    dbRead(config),
)();
