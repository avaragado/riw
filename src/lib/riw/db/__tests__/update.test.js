// @flow

import fs from 'fs';

import mock from 'mock-fs';

import update from '../update';
import cfgBase from '../../__tests__/helpers/dummyConfig';

const cfg = {
    ...cfgBase,
    translationsDatabaseFile: 'db.json',
};

type Fixture = {
    name: string,
    in: RIWDB,
    opt: RIWCLIOptDBUpdate,
};

const fixtures: Fixture[] = [
    {
        name: '01 empty db, no messages to update',
        in: {
            version: 1,
            data: {},
        },
        opt: {
            translations: [],
        },
    },

    {
        name: '02 adding new message',
        in: {
            version: 1,
            data: {},
        },
        opt: {
            translations: [
                {
                    defaultMessage: 'hello',
                    description: 'desc1',
                    locale: 'en-re',
                    translation: 'olleh',
                },
            ],
        },
    },

    {
        name: '03 updating message, same locale',
        in: {
            version: 1,
            data: {
                'hello': {
                    'desc1': {
                        'en-re': 'OLD olleh',
                    },
                },
            },
        },
        opt: {
            translations: [
                {
                    defaultMessage: 'hello',
                    description: 'desc1',
                    locale: 'en-re',
                    translation: 'NEW olleh',
                },
            ],
        },
    },

    {
        name: '04 updating message, new locale',
        in: {
            version: 1,
            data: {
                'hello': {
                    'desc1': {
                        'en-re': 'olleh',
                    },
                },
            },
        },
        opt: {
            translations: [
                {
                    defaultMessage: 'hello',
                    description: 'desc1',
                    locale: 'en-UPPER',
                    translation: 'HELLO',
                },
            ],
        },
    },

    {
        name: '05 several messages at once',
        in: {
            version: 1,
            data: {
                'hello': {
                    'desc1': {
                        'en-re': 'olleh',
                    },
                },
            },
        },
        opt: {
            translations: [
                {
                    defaultMessage: 'hello',
                    description: 'desc1',
                    locale: 'en-UPPER',
                    translation: 'HELLO',
                },
                {
                    defaultMessage: 'goodbye',
                    description: 'desc1',
                    locale: 'en-re',
                    translation: 'eybdoog',
                },
                {
                    defaultMessage: 'foo',
                    description: 'desc1',
                    locale: 'en-re',
                    translation: 'oof',
                },
                {
                    defaultMessage: 'foo',
                    description: 'desc2',
                    locale: 'en-UPPER',
                    translation: 'FOO',
                },
            ],
        },
    },
];

describe('lib/riw/db/update', () => {
    fixtures.forEach((fixture) => {
        it(fixture.name, () => {
            mock({
                [cfg.translationsDatabaseFile]: JSON.stringify(fixture.in),
            });

            update(cfg)(fixture.opt);

            const dbOut: RIWDB = JSON.parse(
                fs.readFileSync(cfg.translationsDatabaseFile).toString(),
            );

            mock.restore();

            expect(dbOut).toMatchSnapshot();
        });
    });
});
