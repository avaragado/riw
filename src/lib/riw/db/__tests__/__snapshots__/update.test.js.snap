// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lib/riw/db/update 01 empty db, no messages to update 1`] = `
Object {
  "data": Object {},
  "version": 1,
}
`;

exports[`lib/riw/db/update 02 adding new message 1`] = `
Object {
  "data": Object {
    "hello": Object {
      "desc1": Object {
        "en-re": "olleh",
      },
    },
  },
  "version": 1,
}
`;

exports[`lib/riw/db/update 03 adding new message with object description 1`] = `
Object {
  "data": Object {
    "hello": Object {
      "{\\"foo\\":123,\\"label\\":\\"desc1\\"}": Object {
        "en-re": "olleh",
      },
    },
  },
  "version": 1,
}
`;

exports[`lib/riw/db/update 04 updating message, same locale 1`] = `
Object {
  "data": Object {
    "hello": Object {
      "desc1": Object {
        "en-re": "NEW olleh",
      },
    },
  },
  "version": 1,
}
`;

exports[`lib/riw/db/update 05 updating message, new locale 1`] = `
Object {
  "data": Object {
    "hello": Object {
      "desc1": Object {
        "en-UPPER": "HELLO",
        "en-re": "olleh",
      },
    },
  },
  "version": 1,
}
`;

exports[`lib/riw/db/update 06 several messages at once 1`] = `
Object {
  "data": Object {
    "foo": Object {
      "desc1": Object {
        "en-re": "oof",
      },
      "{\\"a\\":2,\\"ddd\\":\\"desc2\\",\\"z\\":1}": Object {
        "en-UPPER": "FOO",
      },
    },
    "goodbye": Object {
      "desc1": Object {
        "en-re": "eybdoog",
      },
    },
    "hello": Object {
      "desc1": Object {
        "en-UPPER": "HELLO",
        "en-re": "olleh",
      },
    },
  },
  "version": 1,
}
`;
